common/cuda_hip/preconditioner/batch_ilu_isai.hpp.inc

```diff
@@ -0,0 +1,376 @@
+        __syncthreads();
+        solve_system_relaxation_steps_external_spgemm(
+            num_rows, iter_mat_upper_solve_entry_, u_isai_entry_, y, z, temp);
+    }
+
+public:
+    using value_type = ValueType;
+
+    /**
+     * @param l_batch  Lower Triangular factor (that was generated externally).
+     *
+     * @param u_batch  Upper Triangular factor (that was generated externally).
+     *
+     * @param l_isai_batch  Lower Triangular factor's Incomplete Sparse
+     * Approximate Inverse (that was generated externally).
+     *
+     * @param u_isai_batch  Upper Triangular factor's Incomplete Sparse
+     * Approximate Inverse (that was generated externally).
+     *
+     * @param mult_inv_batch Mutiplication of inverses:  lai_U * lai_U (that was
+     * generated externally).
+     *
+     * @param iter_mat_lower_solve_batch Iteration matrix (= I - lai_L * L ) for
+     * solving L * y = r
+     *
+     * @param iter_mat_upper_solve_batch Iteration matrix (= I - lai_U * U ) for
+     * solving U * z = y
+     *
+     * @param apply_type How the preconditioner is to be applied?
+     *
+     * @param num_relaxation_steps
+     */
+    batch_ilu_isai(
+        const gko::batch_csr::UniformBatch<const value_type>& l_batch,
+        const gko::batch_csr::UniformBatch<const value_type>& u_batch,
+        const gko::batch_csr::UniformBatch<const value_type>& l_isai_batch,
+        const gko::batch_csr::UniformBatch<const value_type>& u_isai_batch,
+        const gko::batch_csr::UniformBatch<const value_type>& mult_inv_batch,
+        const gko::batch_csr::UniformBatch<const value_type>&
+            iter_mat_lower_solve_batch,
+        const gko::batch_csr::UniformBatch<const value_type>&
+            iter_mat_upper_solve_batch,
+        const enum gko::preconditioner::batch_ilu_isai_apply& apply_type,
+        const int& num_relaxation_steps)
+        : l_batch_{l_batch},
+          u_batch_{u_batch},
+          l_isai_batch_{l_isai_batch},
+          u_isai_batch_{u_isai_batch},
+          mult_inv_batch_{mult_inv_batch},
+          iter_mat_lower_solve_batch_{iter_mat_lower_solve_batch},
+          iter_mat_upper_solve_batch_{iter_mat_upper_solve_batch},
+          apply_type_{apply_type},
+          num_relaxation_steps_{num_relaxation_steps}
+    {}
+
+    /**
+     * The size of the work vector required per batch entry. (takes into account
+     * both- generation and application (for application, returns the max. of
+     * what is required by each of the 4 methods))
+     */
+    __host__ __device__ static constexpr int dynamic_work_size(int nrows,
+                                                               int nnz)
+    {
+        return 3 * nrows;
+    }
+
+    /**
+     * Complete the precond generation process.
+     *
+     */
+    __device__ __forceinline__ void generate(
+        size_type batch_id, const gko::batch_csr::BatchEntry<const ValueType>&,
+        ValueType* const __restrict__ work)
```

Comment:
This function signature is same for batch_ilu, batch_ilu_isai and batch_isai because of a common abstracted way to call these functions ?

Reply:
Yeah, that's how is was designed by Aditya. All the internal (device side) batched preconditioner classes, including batch_identity, batch_jacobi, batch_ilu, batch_isai and batch_ilu_isai have this generate function with the same signature, which is called from the batched iterative solver kernels to complete the left-over preconditioner generation process (partial in-solver preconditioner generation has some performance advantages).